docs/proposals/201909_bucketfile.md

```diff
@@ -0,0 +1,329 @@
+if no error occurs, BucketFile will close the `pendingChan` channel.
+
+If callers receive an error from the channel, then prefetch failed. 
+If callers receive nil (zero value of error, after channel closed) from the channel, then prefetch done. 
+
+Each time to fetch pending pages，BucketFile will reset  `pendingPages`, `pendingReaders` and `pendingChan`.
+
+### In-memory object
+
+Each IndexReader decodes [postings offset block](https://github.com/prometheus/prometheus/blob/master/tsdb/docs/format/index.md#postings-offset-table) and [label offset block](https://github.com/prometheus/prometheus/blob/master/tsdb/docs/format/index.md#label-offset-table) into memory object for sequential and  random accesses.  
+
+Each BucketFile has a `pages` bitmap, but the bitmap uses very small memory. 
+
+For inactive IndexReaders, if the inactive duration (`now.Sub(lastQueryTime)`) greats than a configured duration (e.g. 15 mins), then IndexReader will remove in-memory objects and close file descriptor and mmap in IndexFile.  This will reduce memory usage.
+
+**NOTE**: The `pages` bitmap in BucketFile should always be kept in memory until the server is closed.
+
+### Preload blocks on startup
+
+The latest blocks will be queried mostly. 
+
+Preload `the last N hours` (configurable) blocks on startup.
+This will fetch index files from object storage and decode bytes into in-memory objects.
+
+### Local Files
+
+Because the max size of a index file is 64 GiB, and  the max size of a chunks file is 128 MiB.
+
+So we can set the size of a local index file to 64 GiB, set  the size of a local chunks file to 128 MiB
+
+The size is used to create a sparse file and mmap a file.
+
+### Object storage byte ranges
+
+Currently, different provider clients implement byte ranges  differently. We need to wrap their implementations and export the same api.
+
+ I think byte ranges in [minio-go](https://github.com/minio/minio-go/blob/master/api-get-options.go#L99-L128) is a good example:
+
+1. Fetch last N bytes:     
+     `Range{start:0, end: -N}` ->  Request header`Range: bytes=-N`     
+     **NOTE**: will not write to local file because we do not know the offset.
+2. Fetch everything starting from `start`:    
+    `Range{start:N, end: 0}	`  ->  Request heade `Range: bytes=N-` 
+
+3. Fetch everything starting at `start` till  `end`:    
+   `Range{start:N, end: M}` -> Request header `Range: bytes=M-N` 
+
+if providers support multiple ranges, likes `Range: bytes=500-700,60000-61000`, we need to implement a `BucketReader.GetRanges` API, then we can query multiple ranges in one request.
+
+See https://tools.ietf.org/html/rfc7233#section-2.1 for reference.
+
+## Risks
+
+Ensure all providers support `fetch last N bytes` and `fetch everything starting from offset`. Otherwise, we need to know the size of remote index file before fetching TOC block. We may get the size by fetching metadata from object storage.
```

Comment:
Nice! Can we actually check .. now? Design phase is exactly perfect moment for this.

Reply:
providers: Azure, S3, COS, GCS, Swift

all these providers support `fetch last N bytes` and `fetch everything starting from offset`.

some changes in client side:
1. Upgrade GCS client pkg to version v0.45.0+
2. Update Swift [GetRange](https://github.com/thanos-io/thanos/blob/master/pkg/objstore/swift/swift.go#L121) implementation.

https://github.com/googleapis/google-cloud-go/blob/master/CHANGES.md

Reply:
related links:
Azure: https://docs.microsoft.com/en-us/rest/api/storageservices/specifying-the-range-header-for-blob-service-operations

COS: https://cloud.tencent.com/developer/section/1189987

Swift: https://docs.openstack.org/api-ref/object-store/?expanded=get-object-content-and-metadata-detail#get-object-content-and-metadata

S3: https://docs.aws.amazon.com/zh_cn/AmazonS3/latest/API/API_GetObject.html

COS: https://cloud.google.com/storage/docs/xml-api/reference-headers#range

---

COS and Swift support multiple ranges (Range: bytes=A-B,M-N)

Reply:
The `Risks` section has been removed. see https://github.com/thanos-io/thanos/pull/1567/commits/3376912da28e15e5da77386f210e3a1fa1e347c4